### ISSUE TITLE
[Design] Adding Consistency to Labels

### ISSUE THREAD

**Issue by Mornix** [JacquesCarette/Drasil#1022]
Background
=========
`Label`s were recently added (Summer 2018) to alleviate issues with circular references. While the current implementation achieves the original goal, they introduce one sticking point (at least for me). Currently `Label`s eagerly specify their `RefType` which is used to prepend a “grouping” to the `ShortName` when referenced. 

Problem
======
The current implementation introduces two problems:
1. The `Label`’s type is already known because of the chunk type (or related information for`ConceptInstance`) which “owns” the `Label`
2. There is an opportunity to provide inconsistent information (i.e. a `RefType` which is not “the same” as the chunk which "owns" the `Label`)

Solution
======
The solution is to remove `RefType` from `Label`. Simply put, `Label`s in drasil-lang will continue to behave the same way they do currently with chunks. I am a fan of explicit `Label`s being “optional” in chunks. (i.e if a `Label` isn’t solving a circular reference problem, a `ShortName` can be specified to a chunk constructor and a `Label` is generated).

The changes this proposal brings mostly occur in Reference.hs. The `ReferenceDB` will “extract” `Label`s from chunks on insertion. At this point, it is known which chunk type a `Label` is and can be tagged (for deferred `ShortName` generation) or have the `ShortName` updated there. Referencing `RefAdd` (and possibly `MetaLink`) `Label`s directly will no longer pull a `ShortName` from an explicit `Label` but use the provided `UID` to look up the `Label` in the `ReferenceDB`. 

Benefits
======
There are several benefits to this approach:
1. Reducing redundancy in specifying `Label`s
2. `Label`s are now consistent with the chunk they reference
3. By storing `Label`s it is possible to detect if a `Label` is “owned” by more than one chunk
4. No need to explicitly add `Label`s to the `ReferenceDB`
5. Supports migrating the prefixes to drasil-data and exposing user-changeable prefixes
6. Lazy `Label` creation (Only explicitly needed to avoid circular references)
7. Detect orphaned Labels (i.e. `Label`s which _should_ refer to something but points to nothing)

Issues
====
1. Certain chunks are not currently added to the `ReferenceDB`. Things like `TheoryModel`s are only added to the `ChunkDB` and would need to be added to the `ReferenceDB` 
2. (Existing Issue) Currently the `ReferenceDB` is largely superfluous in terms of actually being a necessity to reference most things — It’s more used to gather groups of chunks and allow for referencing by ordered number (but only for `ConceptInstance` and a few other select chunks such as `Goals`)

(Partial) Alternatives
===============
1. Since `Label`s overlap with @halonazhao’s work where she is currently extracting `Label`s to build traceability matrices, a pre-pass framework could be added to traverse the `DocDesc` and extract various pieces of information allowing for the `ReferenceDB` to be generated by extracting information from any chunks which are instances of `HasLabel`. But this alternative requires:
    1. Standardizing on whether `ConceptInstance`’s (as used by `DocDesc`) should “pull” all chunks with a given domain in a database (like `Assumption` does with `ReferenceDB`), or whether all `ConceptInstance`s to display should be enumerated in a list (like how LUR currently behave) in `DocDesc`
    2. Move chunks which exist in the `ReferenceDB` (such as `Citation`’s and `Goal`’s) to `ChunkDB` and consolidate all chunk knowledge in that database rather than spread them between the two
2. Add a table to `ReferenceDB` to specifically hold `Label`s which will pull `Label`s implicitly from specified chunks which exist in the `ReferenceDB` and have a field to the DB constructor to add chunks (and/or just their `Label`s) like `TheoryModel`s which currently do not exist in the database. 
3. Leave everything as is. (Low hanging fruit for an alternative)

(I’ve assigned @halonazhao as well since this design proposal overlaps with her work and she may have opinions on it.)

**Comment by halonazhao** [comment]
I agree that the design to improve `Label`'s consistency is needed.  

But I have a problem where we may have different understanding. The first is that we may just want to keep making references from labels no matter whether the cyclic problem forces us or not. I had a discussion with @JacquesCarette about this at https://github.com/JacquesCarette/Drasil/commit/f1e82d0feaffbfdd14f1ee4d0027db5f3568ba9b#comments.

But I agree that `label` can be created smarter. I think that `Label` eventually may able to be completely atutomatically generated when the chunk it attached to is generated. `Label` only has `UID`, `LblType`, `ShortName`, `RefType`.

Based on current version,
 the label's `UID` is just chunks' `UID` + `"Label"`. 
`RefType` depends on the `Chunks` type. So they are easily being auto filled. 

`LblType` and `ShortName` may need more thoughts. But I have a thought about `ShortName`, What if we go back to use `shortName` not with specific meanings? Instead, using `A1`, `A2` and so on. So when a chunk is generated, a same type of chunk will be generated enumeric IDs for them. `LblType` might need more design.

But if the `label` can be generated along the `chunk` is generated. Many problem mentioned before can be solved.

**Comment by JacquesCarette** [comment]
@Mornix could you try to adjust your proposal to reflect the meeting we had on Monday?  I think what you had was quite close.

I think what we need is that a Chunk be `Referable` when it provide a `RefAddr` and a `ShortName`. When making a **reference**, then we can decorate things (via the Chunk type). To really do this properly will likely require that we name certain things (like sections) be chunks.

Once @halonazhao has given us a list of the calls to `makeRefFromLbl`, we'll better understand the true difficulties.

**Comment by JacquesCarette** [comment]
I'm fairly sure more of this has been fixed, but a thorough look needs to happen (by me).
